### PR TITLE
Issue #3835 - ensure WebSocketSessions are always stopped

### PR DESCRIPTION
**Pull request by lachlan-roberts**
Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

### CODE REVIEW

**Comment by lachlan-roberts** [comment]
#3835

File: jetty-websocket/websocket-common/src/main/java/org/eclipse/jetty/websocket/common/WebSocketSessionListener.java [hunk 1]
```diff
@@ -20,6 +20,8 @@
 
 public interface WebSocketSessionListener
 {
+    void onSessionCreated(WebSocketSession session);
```
Comment (gregw) [inline jetty-websocket/websocket-common/src/main/java/org/eclipse/jetty/websocket/common/WebSocketSessionListener.java]: When adding new methods to an interface, provide a default implementation.  Actually  you can provide default implementation for all these methods now.

File: jetty-websocket/javax-websocket-client-impl/src/main/java/org/eclipse/jetty/websocket/jsr356/ClientContainer.java [hunk 2]
```diff
@@ -524,6 +524,11 @@ public JsrSessionListenerBridge(JsrSessionListener listener)
             this.listener = listener;
         }
 
+        @Override
+        public void onSessionCreated(WebSocketSession session)
```
Comment (gregw) [inline jetty-websocket/javax-websocket-client-impl/src/main/java/org/eclipse/jetty/websocket/jsr356/ClientContainer.java]: Don't add empty implementations, instead have a default implementation

File: jetty-websocket/javax-websocket-server-impl/src/test/java/org/eclipse/jetty/websocket/jsr356/server/SessionTrackingTest.java [hunk 3]
```diff
@@ -96,6 +96,11 @@ public void testAddRemoveSessions() throws Exception
             CountDownLatch closedLatch = new CountDownLatch(2);
             wsServerFactory.addSessionListener(new WebSocketSessionListener()
             {
+                @Override
+                public void onSessionCreated(WebSocketSession session)
```
Comment (gregw) [inline jetty-websocket/javax-websocket-server-impl/src/test/java/org/eclipse/jetty/websocket/jsr356/server/SessionTrackingTest.java]: ditto

File: jetty-websocket/jetty-websocket-tests/src/test/java/org/eclipse/jetty/websocket/tests/ConcurrentConnectTest.java [hunk 4]
```diff
@@ -113,14 +131,36 @@ public void testConcurrentConnect() throws Exception
...
+        assertTrue(serverFactory.getContainedBeans(WebSocketSession.class).isEmpty());
     }
 
-    public static class MyWebSocketServlet extends WebSocketServlet
+    public static class CloseListener implements WebSocketSessionListener
     {
+        public CountDownLatch closeLatch = new CountDownLatch(MAX_CONNECTIONS);
+
+        @Override
+        public void onSessionCreated(WebSocketSession session)
```
Comment (gregw) [inline jetty-websocket/jetty-websocket-tests/src/test/java/org/eclipse/jetty/websocket/tests/ConcurrentConnectTest.java]: ditto

File: jetty-websocket/jetty-websocket-tests/src/test/java/org/eclipse/jetty/websocket/tests/ConcurrentConnectTest.java [hunk 5]
```diff
@@ -113,14 +131,36 @@ public void testConcurrentConnect() throws Exception
...
+        {
+        }
+
+        @Override
+        public void onSessionOpened(WebSocketSession session)
```
Comment (gregw) [inline jetty-websocket/jetty-websocket-tests/src/test/java/org/eclipse/jetty/websocket/tests/ConcurrentConnectTest.java]: ditto

File: jetty-websocket/jetty-websocket-tests/src/test/java/org/eclipse/jetty/websocket/tests/client/BadNetworkTest.java [hunk 6]
```diff
@@ -139,6 +139,11 @@ public void onClosed(Connection connection)
         WebSocketServerFactory wssf = (WebSocketServerFactory)context.getServletContext().getAttribute(WebSocketServletFactory.class.getName());
         wssf.addSessionListener(new WebSocketSessionListener()
         {
+            @Override
+            public void onSessionCreated(WebSocketSession session)
```
Comment (gregw) [inline jetty-websocket/jetty-websocket-tests/src/test/java/org/eclipse/jetty/websocket/tests/client/BadNetworkTest.java]: ditto

File: jetty-websocket/jetty-websocket-tests/src/test/java/org/eclipse/jetty/websocket/tests/client/ClientOpenSessionTracker.java [hunk 7]
```diff
@@ -63,6 +63,11 @@ public void onClosed(Connection connection)
         this.closeConnectionLatch.countDown();
     }
 
+    @Override
+    public void onSessionCreated(WebSocketSession session)
```
Comment (gregw) [inline jetty-websocket/jetty-websocket-tests/src/test/java/org/eclipse/jetty/websocket/tests/client/ClientOpenSessionTracker.java]: OK this is my last ditto.  If there are other examples, you get the idea....

**Pull request merged**